File: src/test/java/funding/startreum/controller/RewardRestControllerTest.java
package funding.startreum.controller;

import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;

// 수정할때 USER Mock으로 추가 필요
@SpringBootTest
@AutoConfigureMockMvc
class RewardRestControllerTest {
/*
    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private JwtUtil jwtUtil;

    @MockitoBean
    private ProjectRepository projectRepository;

    @MockitoBean
    private RewardService rewardService;

    // JWT 토큰
    private String adminToken;                  // 관리자 토큰
    private String sponsorToken;                // 후원자 토큰
    private String beneficiaryToken;            // 수혜자 토큰

    // 테스트용 상수 값
    private final String ADMIN = "admin";
    private final String SPONSER = "sponsor";
    private final String BENEFICIARY = "beneficiary";
    private final Integer REWARD_ID = 1;
    private final Integer PROJECT_ID = 1;
    private final String DESCRIPTION = "이 리워드는 특별한 혜택을 제공합니다.";
    private final String UPDATE_DESCRIPTION = "수정된 리워드 입니다.";
    private final BigDecimal AMOUNT = new BigDecimal("10000");
    private final BigDecimal UPDATE_AMOUNT = new BigDecimal("20000");

    private static final String CREATE_REQUEST_JSON = """
            {
                "projectId": "1",
                "description": "이 리워드는 특별한 혜택을 제공합니다.",
                "amount": "100000"
            }
            """;

    private static final String UPDATE_REQUEST_JSON = """
            {
                "description": "수정된 리워드 입니다.",
                "amount": "1000"
            }
            """;

    private ResultActions performPost(String url, String token, String requestJson) throws Exception {
        return mockMvc.perform(
                post(url)
                        .header("Authorization", "Bearer " + token)
                        .content(requestJson.stripIndent())
                        .contentType(MediaType.APPLICATION_JSON)
        );
    }


    *//**
     * 테스트 시작 전 계정 및 프로젝트를 설정합니다.
     *//*
    @BeforeEach
    void setUp() {
        // 가상의 프로젝트 생성 (프로젝트 ID: 1, 소유자: SPONSOR)
        createVirtualProject(PROJECT_ID, SPONSER);

        // JWT 토큰 생성
        adminToken = jwtUtil.generateAccessToken(ADMIN, "admin@test.com", ADMIN);
        sponsorToken = jwtUtil.generateAccessToken(SPONSER, "sponsor@test.com", SPONSER);
        beneficiaryToken = jwtUtil.generateAccessToken(BENEFICIARY, "beneficiary@test.com", BENEFICIARY);
    }

    private void createVirtualProject(int projectId, String projectOwner) {
        funding.startreum.domain.users.User user = new funding.startreum.domain.users.User();
        user.setName(projectOwner);
        Project project = new Project();
        project.setCreator(user);

        given(projectRepository.findById(projectId))
                .willReturn(Optional.of(project));
    }

    // ========================================
    // 생성 관련 테스트
    // ========================================

    @Test
    @DisplayName("[생성 201] 관리자 계정으로 리워드 생성 시")
    void createTest1() throws Exception {

        RewardResponse response = new RewardResponse(
                REWARD_ID,
                PROJECT_ID,
                DESCRIPTION,
                AMOUNT,
                LocalDateTime.now(),
                LocalDateTime.now()
        );

        given(rewardService.generateNewRewardResponse(any(RewardRequest.class)))
                .willReturn(response);

        ResultActions result = performPost("/api/reward", adminToken, CREATE_REQUEST_JSON);

        result.andExpect(status().isCreated())
                .andExpect(jsonPath("$.status").value("success"))
                .andExpect(jsonPath("$.message").value("리워드 생성에 성공했습니다."))
                .andExpect(jsonPath("$.data.rewardId").value(REWARD_ID))
                .andExpect(jsonPath("$.data.projectId").value(PROJECT_ID))
                .andExpect(jsonPath("$.data.description").value(DESCRIPTION))
                .andExpect(jsonPath("$.data.amount").value(AMOUNT));
    }

    @Test
    @DisplayName("[생성 201] 수혜자 계정으로 리워드 생성 시")
    void createTest2() throws Exception {

        RewardResponse response = new RewardResponse(
                REWARD_ID,
                PROJECT_ID,
                DESCRIPTION,
                AMOUNT,
                LocalDateTime.now(),
                LocalDateTime.now()
        );

        given(rewardService.generateNewRewardResponse(any(RewardRequest.class)))
                .willReturn(response);

        ResultActions result = performPost("/api/reward", beneficiaryToken, CREATE_REQUEST_JSON);

        result.andExpect(status().isCreated())
                .andExpect(jsonPath("$.status").value("success"))
                .andExpect(jsonPath("$.message").value("리워드 생성에 성공했습니다."))
                .andExpect(jsonPath("$.data.rewardId").value(REWARD_ID))
                .andExpect(jsonPath("$.data.projectId").value(PROJECT_ID))
                .andExpect(jsonPath("$.data.description").value(DESCRIPTION))
                .andExpect(jsonPath("$.data.amount").value(AMOUNT));
    }

    @Test
    @DisplayName("[생성 403] 후원자 계정으로 리워드 생성 시 (권한 부족)")
    void createTest3() throws Exception {
        RewardResponse response = new RewardResponse(REWARD_ID, PROJECT_ID, DESCRIPTION, AMOUNT, LocalDateTime.now(), LocalDateTime.now());

        given(rewardService.generateNewRewardResponse(any(RewardRequest.class)))
                .willReturn(response);

        ResultActions result = performPost("/api/reward", sponsorToken, CREATE_REQUEST_JSON);

        result.andExpect(status().isForbidden());
    }

    // ========================================
    // 조회 관련 테스트
    // ========================================


    @Test
    @DisplayName("[조회 200] 존재하는 리워드 검색")
    void getTest1() throws Exception {
        List<RewardResponse> response = List.of(
                new RewardResponse(
                        REWARD_ID,
                        PROJECT_ID,
                        DESCRIPTION,
                        AMOUNT,
                        LocalDateTime.now(),
                        LocalDateTime.now()
                )
        );

        given(rewardService.generateRewardsResponse(PROJECT_ID))
                .willReturn(response);

        ResultActions result = mockMvc.perform(
                get("/api/reward/" + PROJECT_ID)
        );

        result.andExpect(status().isOk())
                .andExpect(jsonPath("$.status").value("success"))
                .andExpect(jsonPath("$.message").value("리워드 조회에 성공했습니다."))
                .andExpect(jsonPath("$.data").isArray())
                .andExpect(jsonPath("$.data.length()").value(1))
                .andExpect(jsonPath("$.data[0].rewardId").value(REWARD_ID))
                .andExpect(jsonPath("$.data[0].projectId").value(PROJECT_ID))
                .andExpect(jsonPath("$.data[0].description").value(DESCRIPTION))
                .andExpect(jsonPath("$.data[0].amount").value(AMOUNT))
                .andExpect(jsonPath("$.data[0].createdAt").exists())
                .andExpect(jsonPath("$.data[0].updatedAt").exists());
    }

    @Test
    @DisplayName("[조회 200] 비어있는 리워드 조회")
    void getReward_NoProject_ReturnsOk() throws Exception {
        List<RewardResponse> response = Collections.emptyList();

        given(rewardService.generateRewardsResponse(PROJECT_ID))
                .willReturn(response);

        ResultActions result = mockMvc.perform(
                get("/api/reward/" + PROJECT_ID)
        );

        result.andExpect(status().isOk())
                .andExpect(jsonPath("$.status").value("success"))
                .andExpect(jsonPath("$.message").value("리워드가 존재하지 않습니다."))
                .andExpect(jsonPath("$.data").isArray())
                .andExpect(jsonPath("$.data.length()").value(0));
    }*/

//      TODO 수정필요
//    @Test
//    @DisplayName("[수정 200] 수혜자 계정으로 리워드 수정")
//    void updateReward_ReturnsOk() throws Exception {
//        LocalDateTime now = LocalDateTime.now();
//        RewardResponse expectedResponse = new RewardResponse(
//                REWARD_ID,
//                PROJECT_ID,
//                UPDATE_DESCRIPTION,
//                UPDATE_AMOUNT,
//                now,
//                now
//        );
//
//        given(rewardService.generateUpdatedRewardResponse(eq(REWARD_ID), any(RewardUpdateRequest.class)))
//                .willReturn(expectedResponse);
//
//        ResultActions result = mockMvc.perform(
//                put("/api/reward/" + REWARD_ID)
//                        .header("Authorization", "Bearer " + beneficiaryToken)
//                        .content(UPDATE_REQUEST_JSON)
//                        .contentType(MediaType.APPLICATION_JSON)
//        );
//
//        result.andExpect(status().isOk())
//                .andExpect(jsonPath("$.status").value("success"))
//                .andExpect(jsonPath("$.message").value("리워드 수정에 성공했습니다."))
//                .andExpect(jsonPath("$.data.rewardId").value(REWARD_ID))
//                .andExpect(jsonPath("$.data.projectId").value(PROJECT_ID))
//                .andExpect(jsonPath("$.data.description").value(UPDATE_DESCRIPTION))
//                .andExpect(jsonPath("$.data.amount").value(UPDATE_AMOUNT))
//                .andExpect(jsonPath("$.data.createdAt").exists())
//                .andExpect(jsonPath("$.data.updatedAt").exists());
//    }

//    @Test
//    @DisplayName("리워드 수정 실패 - description 누락")
//    void updateReward_MissingDescription_ReturnsBadRequest() throws Exception {
//        int rewardId = 10;
//
//        ResultActions result = mockMvc.perform(
//                put("/api/reward/" + rewardId)
//                        .header("Authorization", "Bearer " + token)
//                        .content("""
//                                {
//                                    "amount": "1000"
//                                }
//                                """.stripIndent())
//                        .contentType(new MediaType(MediaType.APPLICATION_JSON, StandardCharsets.UTF_8))
//        ).andExpect(status().isBadRequest());
//    }
//
//    @Test
//    @DisplayName("리워드 수정 실패 - amount 누락")
//    void updateReward_MissingAmount_ReturnsBadRequest() throws Exception {
//        int rewardId = 10;
//
//        ResultActions result = mockMvc.perform(
//                put("/api/reward/" + rewardId)
//                        .header("Authorization", "Bearer " + token)
//                        .content("""
//                                {
//                                    "description": "수정된 리워드 입니다."
//                                }
//                                """.stripIndent())
//                        .contentType(new MediaType(MediaType.APPLICATION_JSON, StandardCharsets.UTF_8))
//        ).andExpect(status().isBadRequest());
//    }
//
//    @Test
//    @DisplayName("리워드 수정 실패 - amount == 0")
//    void updateReward_AmountZero_ReturnsBadRequest() throws Exception {
//        int rewardId = 10;
//
//        ResultActions result = mockMvc.perform(
//                put("/api/reward/" + rewardId)
//                        .header("Authorization", "Bearer " + token)
//                        .content("""
//                                {
//                                    "description": "수정된 리워드 입니다.",
//                                    "amount": "0"
//                                }
//                                """.stripIndent())
//                        .contentType(new MediaType(MediaType.APPLICATION_JSON, StandardCharsets.UTF_8))
//        ).andExpect(status().isBadRequest());
//    }
//
//    @Test
//    void deleteReward() {
//    }
}

File: src/test/java/funding/startreum/domain/reward/controller/RewardRestControllerTest.kt
package funding.startreum.domain.reward.controller

import com.fasterxml.jackson.databind.ObjectMapper
import funding.startreum.common.config.SecurityConfig
import funding.startreum.domain.reward.dto.request.RewardRequest
import funding.startreum.domain.reward.dto.request.RewardUpdateRequest
import funding.startreum.domain.reward.dto.response.RewardResponse
import funding.startreum.domain.reward.service.RewardService
import funding.startreum.domain.users.service.CustomUserDetailsService
import org.hamcrest.Matchers.`is`
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.mockito.BDDMockito.given
import org.mockito.kotlin.any
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.context.annotation.Import
import org.springframework.http.MediaType
import org.springframework.security.test.context.support.WithMockUser
import org.springframework.test.context.bean.override.mockito.MockitoBean
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import java.math.BigDecimal
import java.time.LocalDateTime

/**
 * @see RewardRestController
 */
@Import(SecurityConfig::class)
@WebMvcTest(controllers = [RewardRestController::class])
@AutoConfigureMockMvc
class RewardRestControllerTest {

    @Autowired
    lateinit var mockMvc: MockMvc

    @Autowired
    lateinit var objectMapper: ObjectMapper

    @MockitoBean
    lateinit var rewardService: RewardService

    @MockitoBean
    lateinit var customUserDetailsService: CustomUserDetailsService

    private val BASE_URL = "/api/reward"

    @Nested
    @DisplayName("#1 리워드 생성 API (POST /api/reward)")
    inner class CreateRewardTest {

        @Test
        @DisplayName("1-1) ADMIN 또는 BENEFICIARY 권한을 가진 사용자는 리워드 생성 성공 201")
        @WithMockUser(username = "admin", roles = ["ADMIN"])
        fun createReward_Success() {
            // given
            val request = RewardRequest(
                projectId = 1,
                description = "Test Reward",
                amount = BigDecimal("100")
            )
            val now = LocalDateTime.now()
            val expectedResponse = RewardResponse(
                rewardId = 10,
                projectId = 1,
                description = "Test Reward",
                amount = BigDecimal("100"),
                createdAt = now,
                updatedAt = null
            )
            given(rewardService.generateNewRewardResponse(any())).willReturn(expectedResponse)

            // when & then
            mockMvc.perform(
                post(BASE_URL)
                    .contentType(MediaType.APPLICATION_JSON)
                    .content(objectMapper.writeValueAsString(request))
            )
                .andExpect(status().isCreated)
                .andExpect(jsonPath("$.message", `is`("리워드 생성에 성공했습니다.")))
                .andExpect(jsonPath("$.data.rewardId", `is`(10)))
                .andExpect(jsonPath("$.data.projectId", `is`(1)))
                .andExpect(jsonPath("$.data.description", `is`("Test Reward")))
                .andExpect(jsonPath("$.data.amount", `is`(100)))
        }

        @Test
        @DisplayName("1-2) 권한 없는 사용자는 403 Forbidden 응답을 받는다")
        @WithMockUser(username = "user", roles = ["USER"])
        fun createReward_Forbidden() {
            val request = RewardRequest(
                projectId = 1,
                description = "Test Reward",
                amount = BigDecimal("100")
            )

            mockMvc.perform(
                post(BASE_URL)
                    .contentType(MediaType.APPLICATION_JSON)
                    .content(objectMapper.writeValueAsString(request))
            )
                .andExpect(status().isForbidden)
        }
    }

    @Nested
    @DisplayName("#2 리워드 조회 API (GET /api/reward/{projectId})")
    inner class GetRewardByProjectIdTest {

        @Test
        @DisplayName("2-1) 리워드가 존재할 경우 정상 조회")
        fun getReward_Success() {
            // given
            val projectId = 1
            val now = LocalDateTime.now()
            val rewardList = listOf(
                RewardResponse(10, projectId, "Reward1", BigDecimal("100"), now, null),
                RewardResponse(11, projectId, "Reward2", BigDecimal("200"), now, null)
            )
            given(rewardService.generateRewardsResponse(projectId)).willReturn(rewardList)

            // when & then
            mockMvc.perform(
                get("$BASE_URL/{projectId}", projectId)
                    .contentType(MediaType.APPLICATION_JSON)
            )
                .andExpect(status().isOk)
                .andExpect(jsonPath("$.message", `is`("리워드 조회에 성공했습니다.")))
                .andExpect(jsonPath("$.data[0].rewardId", `is`(10)))
                .andExpect(jsonPath("$.data[1].rewardId", `is`(11)))
        }

        @Test
        @DisplayName("2-2) 리워드가 없을 경우 '리워드가 존재하지 않습니다.' 메시지를 반환")
        fun getReward_NotFound() {
            // given
            val projectId = 1
            val emptyList = emptyList<RewardResponse>()
            given(rewardService.generateRewardsResponse(projectId)).willReturn(emptyList)

            // when & then
            mockMvc.perform(
                get("$BASE_URL/{projectId}", projectId)
                    .contentType(MediaType.APPLICATION_JSON)
            )
                .andExpect(status().isOk)
                .andExpect(jsonPath("$.message", `is`("리워드가 존재하지 않습니다.")))
                .andExpect(jsonPath("$.data").isEmpty)
        }
    }

    @Nested
    @DisplayName("#3 리워드 수정 API (PUT /api/reward/{rewardId})")
    inner class UpdateRewardTest {

        @Test
        @DisplayName("3-1) ADMIN 또는 BENEFICIARY 권한을 가진 사용자는 리워드 수정을 성공한다")
        @WithMockUser(username = "beneficiary", roles = ["BENEFICIARY"])
        fun updateReward_Success() {
            // given
            val rewardId = 10
            val updateRequest = RewardUpdateRequest(
                description = "Updated Reward",
                amount = BigDecimal("150")
            )
            val now = LocalDateTime.now()
            val updatedResponse = RewardResponse(
                rewardId = rewardId,
                projectId = 1,
                description = "Updated Reward",
                amount = BigDecimal("150"),
                createdAt = now,
                updatedAt = now
            )
            given(rewardService.generateUpdatedRewardResponse(rewardId, updateRequest))
                .willReturn(updatedResponse)

            // when & then
            mockMvc.perform(
                put("$BASE_URL/{rewardId}", rewardId)
                    .contentType(MediaType.APPLICATION_JSON)
                    .content(objectMapper.writeValueAsString(updateRequest))
            )
                .andExpect(status().isOk)
                .andExpect(jsonPath("$.message", `is`("리워드 수정에 성공했습니다.")))
                .andExpect(jsonPath("$.data.rewardId", `is`(rewardId)))
                .andExpect(jsonPath("$.data.description", `is`("Updated Reward")))
                .andExpect(jsonPath("$.data.amount", `is`(150)))
        }
    }

    @Nested
    @DisplayName("#4 리워드 삭제 API (DELETE /api/reward/{rewardId})")
    inner class DeleteRewardTest {

        @Test
        @DisplayName("4-1) ADMIN 또는 BENEFICIARY 권한을 가진 사용자는 리워드 삭제에 성공한다")
        @WithMockUser(username = "admin", roles = ["ADMIN"])
        fun deleteReward_Success() {
            val rewardId = 10

            // when & then
            mockMvc.perform(
                delete("$BASE_URL/{rewardId}", rewardId)
                    .contentType(MediaType.APPLICATION_JSON)
            )
                .andExpect(status().isOk)
                .andExpect(jsonPath("$.message", `is`("리워드 삭제에 성공했습니다.")))
        }
    }
}